fix(mme): Changing the hashtable implementations for UE mgmt in amf to maps

shashidhar-patil opened the pull request:
Signed-off-by: shashidhar-patil <shashidhar.patil@wavelabs.ai>

<!--
    Tag your PR title with the components that it touches along with
    the type of change
    E.g. "fix(orc8r): Fix reindexer race condition" or "feat(agw) ..."
-->

## Summary
Earlier: Hashtable implementation was used for UE mgmt in amf. 

Need for change: The reason for the movement is maps are inbuilt libraries and faster as compared to hashtables.They provide better hashing and in turn reduce collisions. 

Code changes:
As a part of this, the following four UE hashtable structures were converted into map implementations.
1. imsi_amf_ue_id_htbl
2. tun11_ue_context_htbl
3. gnb_ue_ngap_id_ue_context_htbl
4. guti_ue_context_htbl

Approach: A generic template structure is implemented for maps. All four maps are created using this generic structure. The generic structure has the basic get, insert and remove operations.

This document gives a brief on design changes: 
[Hastables_to_maps.pdf](https://github.com/magma/magma/files/7480746/Hastables_to_maps.pdf)
 
<!-- Enumerate changes you made and why you made them -->

## Test Plan
- Verified basic sanity with UERANSIM.
- Please find attached logs and pcap snap for basic registration.
- Verified using UT cases.

Logs: (Dated: 09/11/21)
[mme.log](https://github.com/magma/magma/files/7503338/mme.log)

Snapshot of pcap: (Dated: 09/11/21)
![mapPcap](https://user-images.githubusercontent.com/89975652/140893785-1366ecb3-ecdb-41b5-a1de-d76f465738b4.png)

UT test case: (Dated: 09/11/21)
![maptest](https://user-images.githubusercontent.com/89975652/140893824-bd003302-2d15-46a2-8f4c-6378f9697cba.png)



<!--
    How did you test your change? How do you know it works?
    Add supporting screenshots, terminal pastes, etc. as necessary
-->

## Additional Information

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

shashidhar-patil committed:
fix(mme): Changing the hashtable implementations for UE mgmt in amf to maps

Signed-off-by: shashidhar-patil <shashidhar.patil@wavelabs.ai>

ulaskozat commented:
@ardzoht can you steer this PR?

ardzoht commented on lte/gateway/c/core/oai/tasks/amf/amf_map.h:
```diff
@@ -0,0 +1,181 @@
+
```
add magma LICENSE header

shashidhar-patil replied:
This is comment is addressed.

ardzoht commented on lte/gateway/c/core/oai/tasks/amf/amf_map.cpp:
```diff
@@ -0,0 +1,51 @@
+#include "amf_map.h"
```
add magma LICENSE header

shashidhar-patil replied:
This comment is addressed

ardzoht commented on lte/gateway/c/core/oai/test/amf/test_amf_encode_decode.cpp:
```diff
@@ -842,9 +844,86 @@ TEST(test_dnn, test_amf_validate_dnn) {
-  ::testing::InitGoogleTest(&argc, argv);
-  return RUN_ALL_TESTS();
+
+TEST(test_map, test_map) {
```
we should move this test to its own test file, `test_amf_map.cpp`

shashidhar-patil replied:
This change has been done

ardzoht commented on lte/gateway/c/core/oai/tasks/amf/amf_map.h:
```diff
@@ -0,0 +1,181 @@
+
+#include "3gpp_23.003.h"
+
+namespace std {
```
can we move this file to a common folder for reuse? how about ../core/oai/include/map.h and rename `amf_map` to `map`? that will also help us in the future for migrating other usages of hashtable to this

shashidhar-patil replied:
How about moving both, amf_map.h and amf_map.cpp to a new folder "map" in ../core/oai/lib and renaming them to map.h and map.cpp respectively? If not, what would be an appropriate place for amf_map.cpp?

shashidhar-patil replied:
Ignore the above comment. Will move the amf_map.h to ../core/oai/include/map.h

shashidhar-patil replied:
This comment is addressed.

ardzoht commented on lte/gateway/c/core/oai/tasks/amf/amf_map.h:
```diff
@@ -0,0 +1,181 @@
+typedef map_ts_s<uint64_t, uint64_t> map_uint64_ts_t;
+
+// Map Key: guti_m5_t Data: uint64_t;
+typedef map_ts_s<guti_m5_t, uint64_t> obj_map_uint64_ts_t;
```
```suggestion
typedef map_ts_s<guti_m5_t, uint64_t> guti_m5_map_uint64_ts_t;
```

shashidhar-patil replied:
This comment has been addressed.

shashidhar-patil committed:
Addressed review comments

Signed-off-by: shashidhar-patil <shashidhar.patil@wavelabs.ai>

shashidhar-patil commented:
@ardzoht  Addressed the review comments.

Logs: (Dated: 10/11/21)
[mme.log](https://github.com/magma/magma/files/7511328/mme.log)

Snapshot pcap: (Dated: 10/11/21)
![mapPcapReview](https://user-images.githubusercontent.com/89975652/141084513-ac321637-f959-41ae-a078-21e86dfe4f58.png)

ardzoht commented on lte/gateway/c/core/oai/include/map.h:
```diff
@@ -0,0 +1,232 @@
+  }
+};
+}  // namespace std
+namespace magma5g {
```
can we use `magma` namespace here? since this will be reused for hashtables on non-5g code as well

shashidhar-patil replied:
This change has been addressed.

ardzoht commented on lte/gateway/c/core/oai/include/map.h:
```diff
@@ -0,0 +1,232 @@
+    if (umap.empty()) {
+      return MAP_EMPTY;
+    }
+    if (valueP == nullptr) {
```
this condition could be hit if the `valueP` pointer passed is not initialized, why do we return `MAP_BAD_PARAMETER_VALUE` here?

shashidhar-patil replied:
Yes, this condition checks if the valueP is initialized. This is needed as, after fetching the key-value pair we are coping the fetched value into *valueP. If not, can you please suggest a better way of doing this?

ardzoht replied:
but this is done _before_ fetching the key-value pair no? or am i missing something else?

ganeshg87 replied:
Agree. Before fetching the caller should provide valid key-value parameters.
this was a safety check if the caller passes valueP value nullptr unintentionally. below piece code will not crash.
```
    if (search_result != umap.end()) {
      *valueP = search_result->second;
      return MAP_OK;
      }
  ```
 it was our assumptions based on existing hash table validation checks.
 @ardzoht,  let's conclude from your side. we will address this accordingly.

ardzoht commented on lte/gateway/c/core/oai/include/map.h:
```diff
@@ -0,0 +1,232 @@
+
+// Map Declarations:
+// Map- Key: uint64_t , Data: uint64_t
+typedef map_ts_s<uint64_t, uint64_t> map_uint64_ts_t;
```
I think it would be good to start using a definitive naming for the maps, could be `map_<keyType>_<valueType>_ts_t`, in this case (`map_uint64_uint64_ts_t`)

shashidhar-patil replied:
This comment has been addressed.

ardzoht commented on lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp:
```diff
@@ -51,6 +52,7 @@ void amf_ue_context_update_coll_keys(
     const teid_t amf_teid_n11, const guti_m5_t* const guti_p) {
   hashtable_rc_t h_rc                 = HASH_TABLE_OK;
+  map_rc_t m_rc                       = MAP_OK;
   hash_table_ts_t* amf_state_ue_id_ht = get_amf_ue_state();
```
can we add a `TODO` here to note that we want to migrate this hash_table as well?

shashidhar-patil replied:
This comment is addressed.

ardzoht commented on lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp:
```diff
@@ -280,15 +257,13 @@ static bool amf_app_construct_guti(
+  m_rc = amf_ue_context_p->guti_ue_context_htbl.get(*guti_p, &amf_ue_ngap_id64);
 
-  if (HASH_TABLE_OK == h_rc) {
+  if (MAP_OK == m_rc) {
```
```suggestion
  if (m_rc == MAP_OK) {
```

shashidhar-patil replied:
This comment has been addressed.

ardzoht reviewed (changes requested):
minor additional comments left, this is looking good, thanks @shashidhar-patil

shashidhar-patil committed:
Addressed Additional review comments

Signed-off-by: shashidhar-patil <shashidhar.patil@wavelabs.ai>

ulaskozat commented on lte/gateway/c/core/oai/include/map.h:
```diff
@@ -0,0 +1,228 @@
+    if (valueP == nullptr) {
+      return MAP_BAD_PARAMETER_VALUE;
+    }
+    std::lock_guard<std::mutex> lock(mutex_obj);
```
@ardzoht do we need thread safety as each hash map is/should be local to the thread?

ardzoht replied:
That's a good callout @ulaskozat, I think these can be removed following the expectation that the hashtables should be only used by its owner state thread. Given that we will iteratively will migrate these hashtables, we can follow that expectation during the process of the migration. @shashidhar-patil Let's remove the locking and update the name of the struct accordingly.

shashidhar-patil replied:
@ardzoht Okay will do that.

shashidhar-patil replied:
Addressed this comment in the newest commit.

ardzoht commented:
@shashidhar-patil Please resolve conflicts and let's remove locking on the `map` struct since we don't require thread safety given thread ownership of each hashtable

shashidhar-patil committed:
Addressed review comments(Removed locks)

Signed-off-by: shashidhar-patil <shashidhar.patil@wavelabs.ai>

shashidhar-patil commented:
@ardzoht removed locking on the map struct and resolved merge conflicts.

Logs: (Dated: 22/11/21)
[mme.log](https://github.com/magma/magma/files/7581597/mme.log)

Snap pcap: (Dated: 22/11/21)
![mapLock](https://user-images.githubusercontent.com/89975652/142871799-1231622b-3c40-4e96-bc3d-572204965fb8.png)

ardzoht reviewed (approved):
lgtm
